sys/linux/socket_netlink_route_sched.txt

```diff
@@ -0,0 +1,391 @@
+# Copyright 2018 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+# AF_NETLINK/NETLINK_ROUTE SCHED support.
+
+include <linux/net.h>
+include <uapi/linux/netlink.h>
+include <uapi/linux/rtnetlink.h>
+include <uapi/linux/netfilter.h>
+include <uapi/linux/can.h>
+include <uapi/linux/pkt_cls.h>
+include <uapi/linux/pkt_sched.h>
+include <uapi/linux/tc_act/tc_bpf.h>
+include <uapi/linux/tc_act/tc_connmark.h>
+include <uapi/linux/tc_ematch/tc_em_cmp.h>
+include <uapi/linux/tc_ematch/tc_em_ipt.h>
+include <uapi/linux/tc_ematch/tc_em_meta.h>
+
+sendmsg$nl_route_sched(fd sock_nl_route, msg ptr[in, msghdr_netlink[netlink_msg_route_sched]], f flags[send_flags])
+
+netlink_msg_route_sched [
+	newqdisc	netlink_msg[RTM_NEWQDISC, tcmsg[AF_UNSPEC], qdisc_policy]
+	delqdisc	netlink_msg[RTM_DELQDISC, tcmsg[AF_UNSPEC], qdisc_policy]
+	getqdisc	netlink_msg[RTM_GETQDISC, tcmsg[AF_UNSPEC], nlattr[TCA_DUMP_INVISIBLE, void]]
+	newtclass	netlink_msg[RTM_NEWTCLASS, tcmsg[AF_UNSPEC], tclass_policy]
+	deltclass	netlink_msg[RTM_DELTCLASS, tcmsg[AF_UNSPEC], tclass_policy]
+	gettclass	netlink_msg[RTM_GETTCLASS, tcmsg[AF_UNSPEC], void]
+	newtfilter	netlink_msg[RTM_NEWTFILTER, tcmsg[AF_UNSPEC], filter_policy]
+	deltfilter	netlink_msg[RTM_DELTFILTER, tcmsg[AF_UNSPEC], filter_policy]
+	gettfilter	netlink_msg[RTM_GETTFILTER, tcmsg[AF_UNSPEC], nlattr[TCA_CHAIN, int32]]
+	newtaction	netlink_msg[RTM_NEWACTION, tcamsg[AF_UNSPEC], nlattr[TCA_ACT_TAB, tca_actions]]
+	deltaction	netlink_msg[RTM_DELACTION, tcamsg[AF_UNSPEC], action_gd_policy]
+	gettaction	netlink_msg[RTM_GETACTION, tcamsg[AF_UNSPEC], action_dump_policy]
+] [varlen]
+
+type tcmsg[FAMILY] {
+	family		const[FAMILY, int8]
+	ifindex		ifindex
+	tcm_handle	tcm_handle
+	tcm_parent	tcm_handle
+	tcm_info	tcm_handle
+}
+
+tcm_handle {
+	major	int16
```

Comment:
Oh, how about define tcm_handle as
```
tcm_handle {
        minor   flag[tcm_handle_offsets, int32:16]
        major   flag[tcm_handle_offsets, int32:16]
}
tcm_handle_offsets = 0, 1, 2, 3, 4, 10, 11, 12, 13, 14, 20, 21, 22, 23, 24, 0xfff0, 0xfff1, 0xfff2, 0xfff3, 0xfff4, 0xffff
```
0:4 is a little small as in normal use, people like to add more handles like [this example](http://lartc.org/lartc.html#AEN882) did. So I added more offsets.

Reply:
Sounds good, but do we need 0xfff0, 0xfff2, 0xfff3, 0xfff4? I think we only need 0xffff and 0xfff1 and 0 to construct all special values, no?
Do we need 10 and 20, or can we number them just 0 .. N? That would be simpler, and does not produce impression that these 10 and 20 actually mean something.
Also int32:16 is unnecessary, int16 is fine. int32:16 will only produce clumsy bitfield manipulation code in reproducers.

Reply:
OK, then how about
```
tcm_handle {
        minor   flags[tcm_handle_offsets, int16]
        major   flags[tcm_handle_offsets, int16]
}
tcm_handle_offsets = 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 0xffe0, 0xfff1, 0xfff2, 0xfff3, 0xffff
```
I added 0xffe0, 0xfff1, 0xfff2, 0xfff3 because 
```
#define TC_H_UNSPEC     (0U)
#define TC_H_ROOT       (0xFFFFFFFFU)
#define TC_H_INGRESS    (0xFFFFFFF1U)
#define TC_H_CLSACT     TC_H_INGRESS

#define TC_H_MIN_PRIORITY       0xFFE0U
#define TC_H_MIN_INGRESS        0xFFF2U
#define TC_H_MIN_EGRESS         0xFFF3U
```
`TC_H_MIN_PRIORITY`, `TC_H_MIN_INGRESS` and `TC_H_MIN_EGRESS` are only used for minor part. But I think there is no need to add a new tcm_handle_minor_offsets to handle the minor part. What do you think?